### PR TITLE
[5.3] Add length check when using Bcrypt to hash a password

### PR DESCRIPTION
**Pull request by michaeldyrynda**
This PR addresses an issue where [Bcrypt uses only the first 72 characters](https://en.wikipedia.org/wiki/Bcrypt#User_input) of a string when computing a hash.

This pull request adds a check for the length of the passed in `$value` and throws a `LengthException` if it exceeds 72 character, unless the developer chooses to ignore the check, by passing the `ignore_password_length` option to the `Hash::make()` method.

A PHP [RFC](https://wiki.php.net/rfc/password_hash_spec) was also opened quite some time ago to address this behaviour, but doesn't seem to have moved at all. I think, in that case, that the framework is a good place to check for this behaviour. That seems to be the [thinking](https://marc.info/?l=php-internals&m=140577067811277&w=2) by on of the contributors.

My initial thought was to amend the validation of passwords (in the `AuthController`), adding a `max:72` rule but that wouldn't address any users calling the `Hash` service directly, such as when implementing peppering. Given that the framework currently only implements the `BcryptHasher` out of the box, it may make sense that the default validation includes a `max:72` rule for the `password` field. I'll submit a PR for this to the main Laravel application repo.

``` php
// This test would fail, because $hasher->check() will compare
// the computed hash of the truncated strings, which match.
public function testLongPasswordHashingIsTruncatedWithBcrypt()
{
    $hasher = new Illuminate\Hashing\BcryptHasher;
    $hash = $hasher->make('ThisIsAReallyLongPasswordStringThatWillTriggerALengthExceptionWhenItIsUse|**********************************');
    $pass = 'ThisIsAReallyLongPasswordStringThatWillTriggerALengthExceptionWhenItIsUse|###############################';
    $this->assertFalse($hasher->check($pass, $hash));
}
```

This is the resulting verification when using the `Hash::make()` and `Hash::check()` methods. The `|` character is used as a marker for 72 characters to show that the rest of the string is ignored when computing the hash.

```
>>> Hash::make('ThisIsAReallyLongPasswordStringThatWillTriggerALengthExceptionWhenItIsUse|**********************************');
=> "$2y$10$BDD6vdzyXZTPNTuRb3H4iuuysXuKeZySgp0pklXtkXa4X60H5iPLS"
>>> Hash::check('ThisIsAReallyLongPasswordStringThatWillTriggerALengthExceptionWhenItIsUse|###############################', '$2y$10$au/r5waHW9q.U0Ba3RecHuQlLAXoA7gw80FlXD2n2W1y2ipn8D13S');
=> true
```

I've targeted this PR for the next release, as it will be a (minor?) BC break for any current implementations of the framework, given that a `LengthException` will now be thrown for long strings.

Note that there is no length check in the `check` function, which means that any users who already have a long password will be unaffected, although they will not be able to create a password that exceeds 72 characters.

### CODE REVIEW

**Comment by vlakoff** [comment]
This bites me too, but from a cryptography point of view it's perfectly fine. This 72 characters "crop" doesn't break anything and of course there is no security issue (as of today :trollface:).

Only caveat, an attacker can easily guess by cropping its password we're using bcrypt. But don't worry about this.

You may want to read [this question](http://stackoverflow.com/questions/16594613/how-to-hash-long-passwords-72-characters-with-blowfish) on stackoverflow and the below answer by ircmaxell.

**Comment by vlakoff** [comment]
To avoid cropping the user password, you may be tempted to sha-512 the password before passing it to crypt, [**but don't**](http://blog.ircmaxell.com/2015/03/security-issue-combining-bcrypt-with.html).

**Comment by deringer** [comment]
Hey @vlakoff - thanks for your input. I realise this issue has been closed now, but still interested in discussing further (timezones make it tricky to do so in a timely fashion).

My thoughts behind this had less to do with security of the hashed input, and was more about the user (developer) experience perspective. Given that PHP doesn't appear to be addressing the silent truncation of input, it makes sense to do it in user land rather than not at all.

It seems odd that near enough is good enough when performing hash checks. As long as the first part of the string is correct, we don't need to care about the rest of it? Whilst that my be cryptographically sound, I think there ought to be some awareness around it.

If we're going to hide bcrypt behind a hashing service, it's not a big leap to think that the developer using it won't know about this limitation, so at the very least let them know about it i.e. laravel/docs#2233 and perhaps even enforcing a limit. The PHP docs carry this same notice for both [crypt](http://php.net/crypt) and [password_hash](http://php.net/password_hash).

**Comment by vlakoff** [comment]
I know, this truncation annoys me too, but we have to live with it.

If you really, *really* want to hide the truncation to the end user:

    // "best" pick
    $hashed = base64_encode(hash('sha512', $password, true));

    // simpler but less entropy
    $hashed = hash('sha512', $password);

    // DO NOT DO THIS
    $hashed = hash('sha512', $password, true);

Nevertheless, you should do this in your own app. Introducing any of that into the framework would surely lead some people to make mistakes.

TL;DR: Better sweep this under the rug. You can still tweak it on your side, at your own risk.
